# DeepSME

*De Novo* Non-Canonical Nanopore Basecalling Unlocks Private Communication using Heavily-modified DNA Data at Single-Molecule Level

---

In short, this basecaller model can basecall the fully 5hmC modified DNA sequence with high accuracy.

## Install

### Clone the repository

``` bash
git clone https://github.com/sparkcyf/DeepSME.git
```

### Prepare the conda environment

> [!NOTE]
> You need a CUDA compatible GPU to train and infer the model. We recommend using a GPU with at least 12GB of memory.

> [!NOTE]
> Installation of these conda environments will take about **10** minutes on an x86 server (AMD Ryzen 9 5950X with 128G of RAM and SSD).

The following conda environment may be required to install:

1. **bonito-py38** environment for train and infer the Enhanced Basecaller and Reinforced Basecaller. plase refer to [Document of Bonito](https://github.com/nanoporetech/bonito) to install the environment. We have tested the model on `bonito v0.81` with `python 3.8`, `cuda 11.8`.
2. **tombo-py37** environment for generate the k-mer table for 5hmC modified DNA. plase refer to [Document of Tombo](https://github.com/nanoporetech/tombo) to install the environment.

## Model Weights and Sequencing Datasets

### Model Weights
You can download the 5hmC DeepSME basecaller model (in bonito/dorado format), including configuration and weights, from [Zenodo](https://doi.org/10.5281/zenodo.12704171). After downloading, extract the model using:
```
tar -xvf deepsme_dna_5hmc_r9.4.1_e8_sup_bonito.tar.gz
```

### Sequencing Data
* **Raw Sequencing pod5**: The sequencing pod5 files (raw current signals) used for:
  - 6-mer QC and Reinforced 5hmC DeepSME training (5hmC modified Genomic DNA)
  - DNA Storage Datasets decoding

  are also available from [Zenodo](https://doi.org/10.5281/zenodo.12704171). All pod5 data are generated by Oxford Nanopore Technology (ONT) [FLO-MIN106](https://nanoporetech.com/support/flow-cells/MinION-and-GridION/is-my-flow-cell-flo-min106-or-flo-min114) flowcell and [LSK-110](https://nanoporetech.com/document/genomic-dna-by-ligation-sqk-lsk110) chemistry kit.

  

* **Basecalled FASTQ**: The corresponding basecalled FASTQ files via reinforced 5hmC DeepSME can be downloaded from [NCBI BioProject PRJNA1238011](https://www.ncbi.nlm.nih.gov/bioproject/PRJNA1238011).

## To Reproduce DNA Storage Decode Result

The decode scripts for 5hmC-modified DNA Storage Datasets is available at https://github.com/sparkcyf/DeepSME_DNA_Storage_Decode_scripts. You can follow the instructions there to decode the DeepSME basecalled fastq to binary text and image.

## Usage of 5hmC DeepSME Basecaller

### Basecalling

> [!TIP]
> The model architecture and weight of DeepSME is compatible with Bonito. You may use bonito or bonito-compatible basecaller to basecall the sequence current.

#### Aligned (Output BAM)
``` shell
# Bonito
bonito basecaller \
--chunksize 3600 \
--device cuda:0 \
--reference reference.fasta \
deepsme_dna_5hmc_r9.4.1_e8_sup_bonito/ \
fast5_or_pod5_reads/ > basecalling.bam

# Dorado, only accept POD5
dorado basecaller \
--mm2-opts "-x map-ont" \
--no-trim \
--device cuda:0 \
--reference reference.fasta \
deepsme_dna_5hmc_r9.4.1_e8_sup_dorado/ \
pod5_reads/ > basecalling.bam
```

#### Unaligned (Output fastq)
``` shell
# Bonito
bonito basecaller \
--chunksize 3600 \
--device cuda:0 \
deepsme_dna_5hmc_r9.4.1_e8_sup_bonito/ \
fast5_or_pod5_reads/ > basecalling.fastq

# Dorado, only accept POD5
dorado basecaller \
--mm2-opts "-x map-ont" \
--no-trim \
--device cuda:0 \
--emit-fastq \
deepsme_dna_5hmc_r9.4.1_e8_sup_dorado/ \
pod5_reads/ > basecalling.fastq
```

#### (Optional) K-mer model extraction use uncalled4 based on 5hmC-modified 6-mer model

``` shell
uncalled4 train reference.fasta \
gDNA_fast5/ \
--bam-in basecalling.bam \
-m dna_r9.4.1_450bps_6mer_5hmc.npz \
-p 16 -k 6 --kmer-shift 3 --out-dir kmer_5hmC.k6 \
--train-iterations 5 --init-mode moves
```

## Citation

If you find DeepSME useful in your research, please cite:

``` bibtex
@article {Fan2024.08.15.606762,
	author = {Fan, Qingyuan and Zhao, Xuyang and Li, Junyao and Liu, Ronghui and Liu, Ming and Long, Yanping and Fu, Yang and Feng, Qishun and Zhai, Jixian and Pan, Qing and Li, Yi},
	title = {De Novo Non-Canonical Nanopore Basecalling Unlocks Private Communication using Heavily-modified DNA Data at Single-Molecule Level},
	elocation-id = {2024.08.15.606762},
	year = {2024},
	doi = {10.1101/2024.08.15.606762},
	publisher = {Cold Spring Harbor Laboratory},
	URL = {https://www.biorxiv.org/content/early/2024/08/17/2024.08.15.606762},
	eprint = {https://www.biorxiv.org/content/early/2024/08/17/2024.08.15.606762.full.pdf},
	journal = {bioRxiv}
}
```

